src/custom_types/query.py

```diff
@@ -0,0 +1,47 @@
+from typing import Any
+from datetime import datetime
+from pydantic import BaseModel
+from .location_data_types import Date
+
+
+class QueryLocation(BaseModel):
+    """A query location, e.g.,
+    {
+        "lat": 48,
+        "lon": 11,
+    }.
+    """
+
+    lat: int
+    lon: int
+
+    class Config:
+        frozen = True
+
+    def slug(self, verbose: bool = False) -> str:
+        str_ = f"{abs(self.lat):.2f}" if verbose else f"{abs(self.lat):02}"
+        str_ += "S" if self.lat < 0 else "N"
+        str_ += "_" if verbose else ""
+        str_ += f"{abs(self.lon):.2f}" if verbose else f"{abs(self.lon):03}"
+        return str_ + ("W" if self.lon < 0 else "E")
+
+
+class Query(BaseModel):
+    """A query, e.g.,
+    {
+        "from_date": "20210927",
+        "to_date": "20211015",
+        "query_location": <class 'QueryLocation'>,
+    }.
+    """
+
+    from_date: Date
+    to_date: Date
+    location: QueryLocation
+
+    def to_json(self) -> dict[str, Any]:
+        return {
+            "location": self.location.slug(),
+            "from_date": self.from_date,
+            "to_date": self.to_date,
+        }
```

Comment:
Generating the slug in here is very nice! But every pydantic model has the built-in function `.to_dict()`. Could you rename this to something more explicit?

Reply:
Good call (could also be misleading as location is not `{lat:..., lon:...} `but, e.g., `48N012E`). If we leave it like this (only used for the reports) maybe something like `.json_slug()`?

Reply:
Maybe `.to_slugged_json()`?